### PR TITLE
Improve base padding rule

### LINKED ISSUE #50: Allow configurable test file pattern
## Description

The plugin currently identifies a Jest file by checking the filepath for the string `"test"` or `"spec"`. This probably covers most real world usage, but it could also result in some false positives or false negatives.

**False Positives**

With false positives the impact is low. It's unlikely that the false positive is *also* using functions matching Jest names so it will probably just result in wasted time during a lint run.

- `src/inspectors/RequestInspector.js`
- `src/contestants/ContestantProfile.jsx`
- `test/utils/justSomeUtilFile.js`

**False Negatives**

False negatives will be high impact, because they're the target of the plugin and they're missed.

- `src/components/Thing/ThingTest.jsx`

  Case issue + some people keep the test next to the component and not in a test directory.

- `src/components/Thing/__pruebas__/thing.js`

   Someone has customized their Jest set up.

## Proposal

Jest supports a configuration option called [testRegex](https://jestjs.io/docs/en/configuration.html#testregex-string-array-string) that let's one customize how test files are matched.

It makes sense for this plugin to match its default to the Jest default. It will also be important to allow configuration of the `testRegex` in this plugin in case any users have configured their Jest to something different.

~A possibly clever or possibly terrible solution could be to see if we could read the Jest config instead of duplicating the configuration ability.~ <= Nah... too brittle.

### LINKED ISSUE #51: Fix some weirdness with expression => expect => expression => expect
## Description

If there are multiple expects intermixed with expressions in a test block things can get a little weird.

Example:

```js
test('it does something', () => {
  expressionA();
  expect(true).toEqual(true);
  expressionB();
  expressionC();
  expect(true).toEqual(true);
})
```

Becomes...

```js
test('it does something', () => {
  expressionA();

  expect(true).toEqual(true);
  expressionB();
  expressionC();

  expect(true).toEqual(true);
})
```

When ideally it would be...

```js
test('it does something', () => {
  expressionA();

  expect(true).toEqual(true);

  expressionB();
  expressionC();

  expect(true).toEqual(true);
})
```

## Proposal
I think the fix might just simply be changing the config for `padding-before-expect` in `index.ts` to...

```js
[
    { blankLine: 'always', prev: '*', next: 'expect' },
    { blankLine: 'always', prev: 'expect', next: '*' },
    { blankLine: 'any', prev: 'expect', next: 'expect' },
]
```

As well as adding the same addition to the `padding-before-all` rule.

If my guess is correct, then the fix will just be tossing that in and writing some specs. I could be wrong, though... I haven't tested the suggestion at all.

--------------------
 Update, we are going to update all the rules to match this pattern of `around`

### PR DESCRIPTION
**Pull request by benkimpel**
- Move AST utility functions to separate file
- Create/export enums to help build/read configuration objects
- Add an interface and better property names for configuration objects
- Rename `makeRule` to `createRule` and move to main file
- Delete `src/utils.ts`
- Remove isTestFile and use `overrides.files` to specify lint targets
- Fix last implicit any and enable `noImplicitAny` in tsconfig
- Add a padding-around-all rule built from the others' configs
- Add missing x*/f* Jest functions

Closes #50
Closes #51

### CODE REVIEW

**Comment by benkimpel** [comment]
Ok, _now_ it's ready for a look, @dangreenisrael.

Sorry for the large commit. Given there was a refactor alongside additions, it was too hard to break this into several smaller commits.

I'd recommend skipping the diff on `src/rules/padding.ts` and just reading through the file as a whole. I added a huge comment at the bottom to explain the algorithm, moved a lot of functions to module scope, renamed functions and parameterss, added comments and types for clarity, etc.

File: src/rules/padding.ts [hunk 1]
```diff
@@ -1,115 +1,142 @@
 /**
- * @fileoverview Rule to require or disallow newlines between jest functions
- *   Based on eslint/padding-line-between-statements by Toru Nagashima
- *   See: https://github.com/eslint/eslint/blob/master/lib/rules/padding-line-between-statements.js
+ * Require/fix newlines between jest functions
```
Comment (benkimpel) [inline src/rules/padding.ts]: s/between jest functions/around jest functions/

File: tsconfig.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     "module": "commonjs",
-    "noImplicitAny": false,
+    "noImplicitAny": true,
```
Comment (dangreenisrael) [inline tsconfig.json]: nice!

File: src/ast-utils.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import { AST, SourceCode } from 'eslint';
+import { Node } from 'estree';
+
+export const isSemicolonToken = (token: AST.Token): boolean =>
+  token.value === ';' && token.type === 'Punctuator';
+
+export const isTokenOnSameLine = (
```
Comment (dangreenisrael) [inline src/ast-utils.ts]: Any interest in writing some unit tests for these utils. It isn't necessary as we have integration tests for all the rules - just a nice to have.

Comment (dangreenisrael) [inline src/ast-utils.ts]:
```suggestion
export const areTokensOnSameLine= (
```

File: src/ast-utils.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import { AST, SourceCode } from 'eslint';
+import { Node } from 'estree';
+
+export const isSemicolonToken = (token: AST.Token): boolean =>
```
Comment (dangreenisrael) [inline src/ast-utils.ts]:
```suggestion
export const isTokenASemicolon = (token: AST.Token): boolean =>
```

File: src/index.ts [hunk 5]
```diff
@@ -3,106 +3,260 @@
...
+      prevStatementType: StatementType.Any,
+      nextStatementType: [
+        StatementType.DescribeToken,
+        StatementType.FdescribeToken,
+        StatementType.XdescribeToken,
+      ],
+    },
+    {
+      paddingType: PaddingType.Always,
+      prevStatementType: [
```
Comment (dangreenisrael) [inline src/index.ts]: nice!

File: src/index.ts [hunk 6]
```diff
@@ -3,106 +3,260 @@
...
   recommended: {
     plugins: ['jest-formatting'],
-    rules: {
-      'jest-formatting/padding-around-after-all-blocks': 2,
-      'jest-formatting/padding-around-after-each-blocks': 2,
-      'jest-formatting/padding-around-before-all-blocks': 2,
-      'jest-formatting/padding-around-before-each-blocks': 2,
-      'jest-formatting/padding-around-describe-blocks': 2,
-      'jest-formatting/padding-around-test-blocks': 2,
+    overrides: {
```
Comment (dangreenisrael) [inline src/index.ts]: nice!

File: src/rules/padding.ts [hunk 7]
```diff
@@ -137,203 +164,180 @@ function verifyForAlways(
...
-     * Verify padding lines between the given node and the previous node.
-     */
-    function verify(node): void {
-      const parentType = node.parent.type;
-      const validParent = STATEMENT_LIST_PARENTS.has(parentType);
-
-      if (!validParent) {
-        return;
-      }
+  // This is basically... "eh, whatever"
```
Comment (dangreenisrael) [inline src/rules/padding.ts]: This comment isn't super clear to me

File: src/rules/padding.ts [hunk 8]
```diff
@@ -137,203 +164,180 @@ function verifyForAlways(
...
+ * Verify padding lines between the given node and the previous node.
+ */
+const verifyNode = (node: Node, paddingContext: PaddingContext): void => {
+  const { scopeInfo } = paddingContext;
+
+  // NOTE: ESLint types use ESTree which provides a Node type, however
+  //  ESTree.Node doesn't support the parent property which is added by
+  //  ESLint during traversal. Our best bet is to ignore the property access
+  //  here as it's the only place that it's checked.
+  // @ts-ignore
```
Comment (dangreenisrael) [inline src/rules/padding.ts]: +1 to `@ts-ignore` instead of `any`

File: src/ast-utils.ts (hunk 4 above)
Comment (benkimpel) [reply src/ast-utils.ts]: I like it. I'm on it.

File: src/ast-utils.ts (hunk 3 above)
Comment (benkimpel) [reply src/ast-utils.ts]: :+1:

File: tsconfig.json (hunk 2 above)
Comment (benkimpel) [reply tsconfig.json]: yeah, i'm excited about that one!!

File: src/rules/padding.ts (hunk 7 above)
Comment (benkimpel) [reply src/rules/padding.ts]: Oops. Yes... clear as mud. I'll fix!

**Comment by benkimpel** [comment]
@dangreenisrael I think I've addressed your feedback, but let me know if I've missed anything. I did _some_ work on test coverage for ast utils (ast-utils.spec.ts). Two functions left to cover, but i'd vote to handle them at a later time.

**Comment by dangreenisrael** [comment]
@benkimpel Feel free to merge this

**Comment by benkimpel** [comment]
@dangreenisrael merge coming up! gonna squish the PR feedback commit... not really valuable for historical purposes.

**Pull request merged**